flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java

```diff
@@ -0,0 +1,139 @@
+ * </ul>
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ(1234567890123)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ(1234567890, 0)     // Converts epoch seconds
+ * TO_TIMESTAMP_LTZ(1234567890123, 3)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ('2023-01-01T10:00:00', 'yyyy-MM-dd\'T\'HH:mm:ss') // Parses string using input format
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00', 'yyyy-MM-dd HH:mm:ss', 'UTC') // Parses string using input format and timezone
+ * }</pre>
+ */
+@Internal
+public class ToTimestampLtzFunction extends BuiltInScalarFunction {
+
+    private static final int DEFAULT_PRECISION = 3;
+
+    public ToTimestampLtzFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.TO_TIMESTAMP_LTZ, context);
+    }
+
+    public @Nullable TimestampData eval(Number epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+        if (epoch instanceof Float || epoch instanceof Double) {
+            return DateTimeUtils.toTimestampData(epoch.doubleValue(), precision);
+        }
+        return DateTimeUtils.toTimestampData(epoch.longValue(), precision);
+    }
+
+    public @Nullable TimestampData eval(DecimalData epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Number epoch) {
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public @Nullable TimestampData eval(DecimalData epoch) {
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp) {
+        if (timestamp == null) {
+            return null;
+        }
+
+        return parseTimestampData(timestamp.toString());
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp, StringData format) {
+        if (timestamp == null || format == null) {
+            return null;
+        }
+
+        return parseTimestampData(timestamp.toString(), format.toString());
+    }
+
+    public @Nullable TimestampData eval(
+            StringData dateStr, StringData format, StringData timezone) {
+        if (dateStr == null || format == null || timezone == null) {
+            return null;
+        }
+
+        TimestampData ts = parseTimestampData(dateStr.toString(), format.toString());
+        if (ts == null) {
+            return null;
+        }
+
+        try {
+            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
+            return TimestampData.fromInstant(zoneDate.toInstant());
+        } catch (DateTimeException e) {
+            return null;
+        }
+    }
```

Comment:
As I mentioned earlier (probably missed) Introduce a `parseTimestampData` method with 3 input args in `DateTimeUtils` for that
something like
```java
    public static TimestampData parseTimestampData(String dateStr, int precision, TimeZone timeZone)
            throws DateTimeException {
     
                  return TimestampData.fromInstant(
            fromTemporalAccessor(DateTimeFormatter.ofPattern(format).parse(dateStr), 3)
                    .atZone(ZoneId.of(timezone))
                    .toInstant());
    }
```
+ also be sure that it works for invalid format and invalid timezone

Do we have tests for such invalid inputs?

Reply:
Hi @snuyanzin This method is already in `DateTimeUtils.java`, I decided not to use it because it doesn't handle the exceptions well (eg, when we have invalid **timestamp** and **format**. for those cases, I return null here), and I re-wrote the function in the `ToTimestampLtzFunction.java` class. 

Also, for invalid format and timezone, I have them here: https://github.com/apache/flink/pull/25763/files/982b92f13fbb93b8e0c5244b30d4ae8bbbfa3a25#diff-7588925e682b7f5b4b2e41fdc4992ef0d9f246ce00696f783c36f993b8aeb7aeR950-R973

Reply:
yep, thanks tests are there

about method call
here in this method there should just one call as below (same approach as for other methods in this class)
this is what I mean
let me know whether it is clear or not
```java
    public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }
        // call the method from DateTimeUtils
        parseTimestampData();
        }
```

Reply:
Hi @snuyanzin , thank you for clarifying!

I understand that, but that method doesn't handle the invalid cases the way we want here. When we have invalid **timestamp** and **format** and **zoneID**, here we want to return NULL, but that method just throws exception. That's why I decided to re-write and handle those invalid cases separately.


That method with the particular signature is never used, and I don't think we need to refer to it as source of truth. What do you think? 
<img width="905" alt="Screenshot 2024-12-20 at 8 31 03 AM" src="https://github.com/user-attachments/assets/9f72231e-12d8-446d-9613-e0bc6a6be8a0" />

Reply:
Gave it a try the way your recommended, I don’t think it’s essentially different from what I have, and it actually requires more code change. 

**You recommended way would need:** 
Step 1: Add this method to `DateTimeUtils.java`. 
```
public static TimestampData parseTimestampData(String dateStr, int precision, TimeZone timeZone)
            throws DateTimeException {
        return TimestampData.fromInstant(
                fromTemporalAccessor(DEFAULT_TIMESTAMP_FORMATTER.parse(dateStr), precision)
                        .atZone(timeZone.toZoneId())
                        .toInstant());
    }

```
Step 2: Call it in the `ToTimestampLtzFunction.java` function:
```
public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }

        try {
            return parseTimestampData(dateStr.toString(), format.toString(), ZoneId.of(timezone.toString()));
        } catch (DateTimeException e) {
            return null;
        }
}
```

`My way: `
In `ToTimestampLtzFunction.java`:
 ```
public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }

        TimestampData ts = parseTimestampData(dateStr.toString(), format.toString());
        if (ts == null) {
            return null;
        }

        try {
            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
            return TimestampData.fromInstant(zoneDate.toInstant());
        } catch (DateTimeException e) {
            return null;
}
```
Again, I don't see a particular benefit in doing it one way over another, but I'm open for your feedback.